### PR TITLE
Merge "mapbox-gl-style-spec" repository into this repository

### LINKED ISSUE #3794: Merge the "mapbox-gl-style-spec" repo into this repo
ref #3656 

# Todo

- move over files
- set up `@mapbox/mapbox-gl-style-spec` npm package (name still under discussion) to be published from this repo
- mark the `mapbox-gl-style-spec` npm package as deprecated
- update all Mapbox projects to use new `@mapbox/mapbox-gl-style-spec` package
- configure `mapbox-gl-style-spec` documentation to be published by `mapbox-gl-js`
- clean up `mapbox-gl-style-spec` S3 resources
- migrate `mapbox-gl-style-spec` issues & PRs
- mark the `mapbox-gl-style-spec` github repo as deprecated
- find a way to minify the style reference without checking in build artifacts

### PR DESCRIPTION
**Pull request by lucaswoj**
This PR merges the "mapbox-gl-style-spec" repository into this repository.

fixes #3794 

## Todo Before Merging

- [x] briefly describe the changes in this PR
- [x] fix eslint failures
- [x] fix unit test failures
- [x] fix test suite failures
- [x] audit all new dependencies to ensure they're all needed
- [x] change default test reporter
- [x] remove style spec from `package.json` & wire GL JS up to bundled style spec version
- [x] integrate `build-style-spec-docs` script into `build-docs` scripts
- [x] eliminate `build-style-spec` script in favor of custom browserify middleware
- [x] add a build test to ensure only the latest minified style spec reference JSON is pulled into the minified bundle
- [x] set up `mapbox-gl-style-spec` package to be published from this repo, rename npm package to use new namespace (“@mapbox/…”), write docs on when/how to publish a new style spec version
- [x] eliminate `build-style-spec-api-docs` script in favor of JSDoc flow, ensure style spec modules don't show up in GL JS API docs
- [x] test all `bin` commands
- [x] rewrite style spec README
- [x] manually test the debug page
- [x] add [bubleify](https://www.npmjs.com/package/bubleify) to style spec's `package.json`
- [x] review all [npm packages that depend on style spec](https://www.npmjs.com/browse/depended/mapbox-gl-style-spec)
- [x] pull in all changes from `mapbox-gl-js`
- [x] pull in all changes from `mapbox-gl-style-spec`
- [x] `update docs/_layouts/docs.html` with new style spec links

## Todo Immediately After Merging

- [ ] publish `@mapbox/mapbox-gl-style-spec` npm package
- [ ] mark the `mapbox-gl-style-spec` npm package as deprecated
- [ ] update all Mapbox projects to use new `@mapbox/mapbox-gl-style-spec` package
- [ ] migrate `mapbox-gl-style-spec` issues & PRs
- [ ] mark the `mapbox-gl-style-spec` github repo as deprecated
- [ ] set up redirects from the old `mapbox-gl-style-spec` website

## Follow-Up PR

- [ ] move all issues from `mapbox-gl` to this repo
- [ ] rename this repo to `mapbox-gl`
- [ ] rename s3 bucket `mapbox-gl`
- [ ] set up redirects from `mapbox-gl-js` website to `mapbox-gl` website

### CODE REVIEW

**Comment by lucaswoj** [comment]
This is ready for review @jfirebaugh @mollymerp @lbud @anandthakker 

I recommend reviewing in these chunks:

 1. eslint fixes galore: https://github.com/mapbox/mapbox-gl-js/commit/12bf40dce40f37feb2b7ba179d89acaed5e45bff
 2. the rest of the changes: https://github.com/mapbox/mapbox-gl-js/compare/71c54e06f0c17dcf0e421a775122c7150f4d9a28...merge-style-spec

I'll need to do some sort of sync of both repo's `HEAD`s before we actually 🚢  this.

**Comment by jfirebaugh** [comment]
>  1. eslint fixes galore: https://github.com/mapbox/mapbox-gl-js/commit/12bf40dce40f37feb2b7ba179d89acaed5e45bff

This is primarily the introduction of ES6 features (`let/const`, arrow functions, template strings, etc.) that we've previously not used in the mapbox-gl-style-spec module. Will we now require users of this module to be running ES6-supporting environments (technically a breaking change)? Or are we introducing a transpile step for the mapbox-gl-style-spec publishing workflow? Does it support users of the module who may be requiring specific .js files?

File: js/ui/camera.js [hunk 1]
```diff
@@ -327,8 +327,8 @@ class Camera extends Evented {
             nw = tr.project(bounds.getNorthWest()),
             se = tr.project(bounds.getSouthEast()),
             size = se.sub(nw),
-            scaleX = (tr.width - options.padding * 2 - Math.abs(offset.x) * 2) / size.x,
```
Comment (jfirebaugh) [inline js/ui/camera.js]: Is this an intended part of this PR?

File: js/style-spec/API.md [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+  var validate = require('mapbox-gl-style-spec').validate;
+  var style = fs.readFileSync('./style.json', 'utf8');
+  var errors = validate(style);
+```
+
+
+**Returns** `Array.<ValidationError|ParsingError>`, 
+
+
+### `validateStyleMin(style, [styleSpec])`
```
Comment (jfirebaugh) [inline js/style-spec/API.md]: Custom browserify approach means we can merge this with `validate`. 👍  if you want to do that in a separate PR.

File: js/ui/camera.js (hunk 1 above)
Comment (lucaswoj) [inline js/ui/camera.js]: It is not. Reverting in a subsequent commit.

File: js/style-spec/API.md (hunk 2 above)
Comment (lucaswoj) [inline js/style-spec/API.md]: :point https://github.com/mapbox/mapbox-gl-js/issues/4056

**Comment by lucaswoj** [comment]
> Will we now require users of this module to be running ES6-supporting environments (technically a breaking change)? Or are we introducing a transpile step for the mapbox-gl-style-spec publishing workflow?

I surmise that most downstream users of `mapbox-gl-style-spec` are using it within `node`, not in the browser.  It should cause minimal disruption to announce that `mapbox-gl-style-spec` now has the same `node` version requirements as GL JS (>= 4.0.0). 

> Does it support users of the module who may be requiring specific .js files?

Users who are  requiring specific .js files will need to remove the `lib` directory from their path. This breaking change is mitigated by

 - the fact the package is changing names, forcing some developer attention and creating a semver grey area (nobody will be automatically upgraded to this version)
 - requiring specific .js files is generally not considered part of package's public API (users expect breakage from time to time)
 - the fix is easy and intuitive

**Comment by jfirebaugh** [comment]
👍  Sounds reasonable to me.

**Comment by anandthakker** [comment]
> I surmise that most downstream users of mapbox-gl-style-spec are using it within node, not in the browser.

@tmcw ^ heads up re: Studio

**Comment by tmcw** [comment]
Yeah, as it stands, this would break Studio with no easy recourse: [there's an existing ticket tracking modules that have gone ES6-without-transpilation](https://github.com/mapbox/studio/issues/8325), and there's no good way to solve the problem in browserify. In webpack-land, where we might go in ~4 months, you can include external modules in babel's transpilation, but with browserify, the bundler (reasonably, I'd say) requires dependencies to either declare their required transforms with the browserify field, or be written in the target language.

**Comment by anandthakker** [comment]
That's what I was afraid of 😞 

It seems to me that the choices here would be:
 - do some ugly, brittle hackery in Studio's build step (like explicitly bundling `node_modules/mapbox-gl/...` w/ the right transpilation transforms)
 - attempt to declare the appropriate browserify info using the `browser` and `browserify` fields of style-spec's `package.json`

The former isn't great for its brittleness.  The latter isn't great because it means maintaining that build metadata over time, but I think that's better than the former?

There is _almost_ a third option: we could just export the style spec off of the `mapboxgl` object, since that would leverage the browserify transform pipeline that's already happening when the main mapbox-gl module is built.  However, mapbox-gl minifies the spec, stripping out the documentation, and I'm pretty sure Studio needs that stuff.

**Comment by tmcw** [comment]
Also, noting that given the [current public dependencies](https://www.npmjs.com/browse/depended/mapbox-gl-style-spec), 6 of the 12 unique public dependents of mapbox-gl-style-spec are browser-facing.

**Comment by anandthakker** [comment]
> There is almost a third option: we could just export the style spec off of the mapboxgl object, since that would leverage the browserify transform pipeline that's already happening when the main mapbox-gl module is built. However, mapbox-gl minifies the spec, stripping out the documentation, and I'm pretty sure Studio needs that stuff.

Hmm... what if we stuck a check for `process.env.MGL_INCLUDE_SPEC_DOCS` in `minify_style_spec.js` and then exported the spec & utils from the `mapboxgl` object?  That would allow Studio and other web-based dependents to do:
 - `MGL_INCLUDE_SPEC_DOCS=1 existing_browserify_command ...`
 - and just change `require('mapbox-gl-style-spec')` to `require('mapbox-gl').styleSpec.{reference,validate,...}`

Not sure if this is actually any better than option 2, though.

**Comment by lucaswoj** [comment]
Thanks for raising these concerns @anandthakker @tmcw. You've brought to light some points I hadn't considered. My plan of attack:

 - add a "browserify" field to style spec's `package.json` which can reliably transform the package into vanilla javascript (done in https://github.com/mapbox/mapbox-gl-js/pull/3946/commits/cc80268968e6d43770d819032a0b0e54ea340e08)
 - review public npm packages that depend on style spec, work with the authors of any which use style spec's utility functions, do not use browserify, and need to be run in the browser

File: js/style-spec/package.json [hunk 3]
```diff
@@ -20,12 +20,28 @@
...
+      [
+        "bubleify",
+        {
+          "bubleError": true,
+          "transforms": {
+            "dangerousForOf": true
+          }
+        }
+      ],
+      "unassertify",
```
Comment (jfirebaugh) [inline js/style-spec/package.json]: Bad trailing comma.

Comment (lucaswoj) [inline js/style-spec/package.json]: Fixed 👍

**Comment by jfirebaugh** [comment]
Integration in gl-native looks good: https://github.com/mapbox/mapbox-gl-native/commit/c9f38506aac2d2ee112f2a9fe23de95a374c317d.
